core/src/main/scala/cats/derived/contravariant.scala

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright (c) 2015 Miles Sabin
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cats
+package derived
+
+import shapeless._
+import util.VersionSpecific.OrElse
+
+import scala.annotation.implicitNotFound
+
+@implicitNotFound("Could not derive an instance of Contravariant[${F}]")
+trait MkContravariant[F[_]] extends Contravariant[F] {
+  def safeContramap[A, B](fa: F[A])(f: B => Eval[A]): Eval[F[B]]
+  def contramap[A, B](fa: F[A])(f: B => A): F[B] = safeContramap(fa)(f andThen Eval.now).value
+}
+object MkContravariant extends MkContravariantDerivation {
+  def apply[F[_]](implicit F: MkContravariant[F]): MkContravariant[F] = F
+}
+private[derived] abstract class MkContravariantDerivation extends MkContravariantNested {
+  implicit val mkContraHNil: MkContravariant[Const[HNil]#λ] = mkContraConst
+  implicit val mkContraCNil: MkContravariant[Const[CNil]#λ] = mkContraConst
+
+  implicit def mkContraConst[T]: MkContravariant[Const[T]#λ] =
+    new MkContravariant[Const[T]#λ] {
+      def safeContramap[A, B](t: T)(f: B => Eval[A]): Eval[T] = Eval.now(t)
+    }
+}
+
+private[derived] abstract class MkContravariantNested extends MkContravariantCons {
+  protected type FunctorOrMk[F[_]] = Functor[F] OrElse MkFunctor[F]
+  protected def mkSafeMap[F[_], A, B](F: FunctorOrMk[F])(fa: F[A])(f: A => Eval[B]): Eval[F[B]] =
+    F.unify match {
+      case mk: MkFunctor[F] => mk.safeMap(fa)(f)
+      case p => Eval.later(p.map(fa)(f(_).value))
+    }
+
+  implicit def mkContravariantFunctorNested[F[_]](implicit F: Split1[F, FunctorOrMk, ContraOrMk]): MkContravariant[F] =
+    new MkContravariant[F] {
+      def safeContramap[A, B](fa: F[A])(f: B => Eval[A]): Eval[F[B]] =
+        mkSafeMap(F.fo)(F.unpack(fa))(
+          (a: F.I[A]) => mkContraSafe(F.fi)(a)(f)
+        ).map(F.pack)
+    }
+}
```

Comment:
Seems like `F[G[H[_]]]` should also be contra when `F`, `G`, and `H` are but my attempts to do something like that blew up a bit.

Reply:
I think here we would hit the limitations of shapeless' `Split1`. I think it would be enough to have two instances for `Contravariant`:
  * `F[_]: Functor, G[_]: ContravariantOrMk => Contravariant[F[G[_]]]`
  * `F[_]: ContravariantOrMk, G[_]: Functor => Contravariant[F[G[_]]]`

And then we might also want to add another one to `MkFunctor` (I think we are missing that):
  * `F[_]: Contravariant, G[_]: Contravariant => Functor[F[G[_]]]`

Reply:
>   `F[_]: Functor, G[_]: ContravariantOrMk => Contravariant[F[G[_]]]`
>   `F[_]: ContravariantOrMk, G[_]: Functor => Contravariant[F[G[_]]]`

Note that those would have to be prioritized but it doesn't matter in which order.

Reply:
>  `F[_]: Contravariant, G[_]: Contravariant => Functor[F[G[_]]]`

in `Functor` you could add this in this PR or could be left for later, your choice.

Reply:
Aha, that makes good sense. That makes a lot more sense than what I tried to do with multiple `Split1` evidences.